pkg/client/event/deduplicator.go

```diff
@@ -0,0 +1,193 @@
+// event is a duplicate and should be ignored or if it is not a duplicate and
+// should be handled.
+//
+// Four events are supported:
+// - key generation request for a new keep,
+// - signature request for a keep,
+// - keep close request,
+// - keep terminate request.
+type Deduplicator struct {
+	keepRegistry keepRegistry
+	chain        chain.Handle
+
+	keyGenKeeps         *keepEventTrack
+	requestedSignatures *requestedSignaturesTrack
+	closingKeeps        *keepEventTrack
+	terminatingKeeps    *keepEventTrack
+}
+
+type keepRegistry interface {
+	HasSigner(keepAddress common.Address) bool
+}
+
+func NewDeduplicator(
+	keepRegistry keepRegistry,
+	chain chain.Handle,
+) *Deduplicator {
+	keyGenKeeps := &keepEventTrack{
+		data: make(map[string]bool),
+	}
+	requestedSignatures := &requestedSignaturesTrack{
+		data: make(map[string]map[string]bool),
+	}
+	closingKeeps := &keepEventTrack{
+		data: make(map[string]bool),
+	}
+	terminatingKeeps := &keepEventTrack{
+		data: make(map[string]bool),
+	}
+
+	return &Deduplicator{
+		keepRegistry:        keepRegistry,
+		chain:               chain,
+		keyGenKeeps:         keyGenKeeps,
+		requestedSignatures: requestedSignatures,
+		closingKeeps:        closingKeeps,
+		terminatingKeeps:    terminatingKeeps,
+	}
+}
+
+// NotifyKeyGenStarted notifies the client wants to start key generation for
+// a keep upon receiving an event. It returns boolean indicating whether the
+// client should proceed with the execution or ignore the event as a duplicate.
+//
+// In case the client proceeds with the key generation, it should call
+// notifyKeyGenCompleted once the protocol completes, no matter if it failed or
+// succeeded.
+func (d *Deduplicator) NotifyKeyGenStarted(keepAddress common.Address) bool {
+	if d.keyGenKeeps.has(keepAddress) {
+		return false
+	}
+
+	if d.keepRegistry.HasSigner(keepAddress) {
+		return false
+	}
+
+	return d.keyGenKeeps.add(keepAddress)
+}
+
+// NotifyKeyGenCompleted should be called once client completed key generation
+// protocol, no matter if it succeeded or not.
+func (d *Deduplicator) NotifyKeyGenCompleted(keepAddress common.Address) {
+	d.keyGenKeeps.remove(keepAddress)
+}
+
+// NotifySigningStarted notifies the client wants to start signature generation
+// for the given keep and digest upon receiving an event. It returns boolean
+// indicating whether the client should proceed with the execution or ignore the
+// event as a duplicate.
+//
+// In case the client proceeds with signing, it should call
+// notifySigningCompleted once the protocol completes, no matter if it failed or
+// succeeded.
+func (d *Deduplicator) NotifySigningStarted(
+	keepAddress common.Address,
+	digest [32]byte,
+) (bool, error) {
+	if d.requestedSignatures.has(keepAddress, digest) {
+		return false, nil
+	}
+
+	// repeat the check in case of a small chain reorg or if chain nodes
+	// are out of sync
+	isAwaitingSignature, err := utils.ConfirmWithTimeout(
```

Comment:
This is something different. This check covers the situation when we received an event, we are not handling this event at the moment (not an immediate duplicate) and we need to validate whether this is something recent or some event from the past. For this reason, we execute on-chain call and we retry this call only to address situations when Ethereum nodes behind a load balancer are not yet in sync. If we can not confirm the signing is pending, we end up with `shouldHandle = false` and we consider it a normal situation.

The referenced check in `client.go` addresses the situation of chain reorg when the event was emitted, signing request was confirmed on-chain but it disappeared from there as a result of a chain reorg and we do not want to proceed with signing because adversary would be able to report fraud.

I do agree these two checks could probably be combined assuming we do not want to log warning for the second one but I am not sure if it would be the right call. The context is different, basically.